osm2streets/src/transform/classify_intersections.rs

```diff
@@ -1,85 +1,229 @@
+    return true;
+}
 
-        // detect the simple case of a single road splitting
-        if num_roads_inout == 1 {
-            // TODO Determine if it is possible to turn from the 'in' road to the 'out' road.
-            // If not, then we have a dual carriageway split.
-            if false {
-                return MultiConnection;
+fn turn_is_allowed(src: &Road, dst: &Road) -> bool {
+    let mut has_exclusive_allows = false;
+    for (t, other) in src.turn_restrictions.iter() {
+        match t {
+            RestrictionType::BanTurns => {
+                if *other == dst.id {
+                    return false;
+                }
+            }
+            RestrictionType::OnlyAllowTurns => {
+                if *other == dst.id {
+                    return true;
+                }
+                has_exclusive_allows = true;
             }
         }
     }
+    !has_exclusive_allows
+}
+
+fn calc_conflict(a: &(usize, usize), b: &(usize, usize), side: DrivingSide) -> ConflictType {
+    use ConflictType::*;
+
+    // If the traffic starts of ends at the same place in the same direction...
+    if a.0 == b.0 && a.1 == b.1 {
+        return Uncontested;
+    }
+    if a.0 == b.0 {
+        return Diverge;
+    }
+    if a.1 == b.1 {
+        return Merge;
+    }
 
-    // A Merge is characterised by an uninterrupted road (that would qualify as a Connection or
-    // MultiConnection), with additional connected roads that yield.
-    // TODO Combine roads into corridors and count them.
+    // The intersection has a boundary that we have labelled 0 to n-1 in clockwise order (from an
+    // arbitrary point), like a string laying in a circle. If we represent `a` as an arc from one
+    // point on the string to another, then there is a section of the string between the two points,
+    // connecting them the two points and two ends of string "on the outside". A second arc, `b`,
+    // crosses `a` if and only if `b` has one end between the points and one end outside.
+    //     ______
+    //    /  |   \
+    //   |   |a   n
+    //   |   |    0
+    //    \__|___/
+
+    // What if the traffic meets going in opposite directions?
+    // It depends on where the traffic came from, and which side we're driving on.
+
+    // Below: If a movement going in the other direction, `b`, joins the indicated LHT movement `a`
+    // (at either end), it will join the road on the dotted side. Whether the other end of `b` is
+    // between the endpoints of `a` or not corresponds to the crossing of the road.
+    // Therefore, if `a` is drawn pointing upwards from low .0 to high .1,
+    // then LHT would be crossed by movements joining from the "inside".
+    //     ______          ______
+    //    /  ^:  \        /  :|  \
+    //   |  a|:   n      |   :|   n
+    //   |   |:   0      |   :|a  0
+    //    \__|:__/        \__:V__/
+
+    // This equation (hopefully) works. Once it does, just trust it:
+    // TODO unit test these three equations.
```

Comment:
The results look pretty good, but I found at least one error, which I think comes from these equations. northgate_dual_carriageway [intersection 95600603](https://a-b-street.github.io/osm2streets/?test=northgate_dual_carriageway#20.97/47.70805/-122.32864) should be a Merge, but is classified as a Cross.

Considering that I'd prefer to work with `HalfRoad`s instead, I might delay unit testing and figuring this out properly.

Reply:
I got kind of lost around here. I'm fine with leaving this alone for now and iterating later, likely with `HalfRoads`.

Also illegal u-turns being generated right now cause some unexpected results in `aurora_sauage_link`, but that'll go away with a fix for that